File: internal/hetzner/ssh_keys.go
package hetzner

import (
	"context"
	"crypto/md5" //nolint:gosec // MD5 is required for Hetzner SSH key fingerprints (RFC 4716)
	"encoding/base64"
	"fmt"
	"strings"

	"github.com/hetznercloud/hcloud-go/v2/hcloud"
)

// EnsureSSHKey ensures the given public key exists in Hetzner.
// Returns the Hetzner SSH key ID.
// If a key with the same fingerprint exists, returns its ID.
// Otherwise, creates a new key and returns its ID.
func (c *Client) EnsureSSHKey(ctx context.Context, name, publicKey string) (string, error) {
	// Calculate fingerprint
	fingerprint, err := calculateFingerprint(publicKey)
	if err != nil {
		return "", fmt.Errorf("failed to calculate fingerprint: %w", err)
	}

	// Check if key already exists by fingerprint
	existingKey, _, err := c.hc.SSHKey.GetByFingerprint(ctx, fingerprint)
	if err != nil {
		return "", fmt.Errorf("failed to check existing SSH keys: %w", err)
	}

	if existingKey != nil {
		return fmt.Sprintf("%d", existingKey.ID), nil
	}

	// Create new SSH key
	opts := hcloud.SSHKeyCreateOpts{
		Name:      name,
		PublicKey: publicKey,
	}

	newKey, _, err := c.hc.SSHKey.Create(ctx, opts)
	if err != nil {
		return "", fmt.Errorf("failed to create SSH key: %w", err)
	}

	return fmt.Sprintf("%d", newKey.ID), nil
}

// GetSSHKeyByID retrieves an SSH key by its ID.
func (c *Client) GetSSHKeyByID(ctx context.Context, id int64) (*hcloud.SSHKey, error) {
	key, _, err := c.hc.SSHKey.GetByID(ctx, id)
	if err != nil {
		return nil, fmt.Errorf("failed to get SSH key: %w", err)
	}
	return key, nil
}

// calculateFingerprint calculates the MD5 fingerprint of an SSH public key.
// Format: xx:xx:xx:... (colon-separated hex pairs)
func calculateFingerprint(publicKey string) (string, error) {
	// Parse the public key to extract the base64 part
	parts := strings.Fields(publicKey)
	if len(parts) < 2 {
		return "", fmt.Errorf("invalid public key format")
	}

	// Decode the base64 part
	keyData, err := base64.StdEncoding.DecodeString(parts[1])
	if err != nil {
		return "", fmt.Errorf("failed to decode public key: %w", err)
	}

	// Calculate MD5 hash (required by Hetzner API for fingerprint matching)
	hash := md5.Sum(keyData) //nolint:gosec // MD5 required for Hetzner SSH key fingerprints

	// Format as colon-separated hex pairs
	var fp strings.Builder
	for i, b := range hash {
		if i > 0 {
			fp.WriteString(":")
		}
		fmt.Fprintf(&fp, "%02x", b)
	}

	return fp.String(), nil
}

File: src/digitalocean/ssh-keys.ts
import { createHash } from "node:crypto";

import type { DigitalOceanSSHKey } from "@/digitalocean/client";

export interface SSHKeyClient {
	listSSHKeys(): Promise<DigitalOceanSSHKey[]>;
	createSSHKey(name: string, publicKey: string): Promise<DigitalOceanSSHKey>;
}

export function calculateFingerprint(publicKey: string): string {
	const parts = publicKey.trim().split(/\s+/);
	if (parts.length < 2) {
		throw new Error("invalid public key format");
	}

	const keyData = Buffer.from(parts[1], "base64");
	if (keyData.length === 0) {
		throw new Error("invalid public key encoding");
	}

	const hex = createHash("md5").update(keyData).digest("hex");
	return hex.match(/.{1,2}/g)?.join(":") ?? "";
}

export async function ensureSSHKey(
	client: SSHKeyClient,
	name: string,
	publicKey: string,
): Promise<string> {
	const fingerprint = calculateFingerprint(publicKey);

	const existing = await client.listSSHKeys();
	const found = existing.find((key) => key.fingerprint === fingerprint);
	if (found) {
		return String(found.id);
	}

	try {
		const created = await client.createSSHKey(name, publicKey);
		return String(created.id);
	} catch (error) {
		const raced = await client.listSSHKeys();
		const raceWinner = raced.find((key) => key.fingerprint === fingerprint);
		if (raceWinner) {
			return String(raceWinner.id);
		}
		throw error;
	}
}
